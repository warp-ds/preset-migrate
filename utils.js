export const CSS_DOCS_URL = "https://warp-ds.github.io/css-docs";
export const TECH_DOCS_URL = "https://warp-ds.github.io/tech-docs";

export const CONSOLE_COLORS = {
  warn: "\x1b[93m",
  error: "\x1b[91m",
  pink: "\x1b[95m",
  bright: "\x1b[1m",
  reset: "\x1b[0m",
};

export const COMING_SOON_MSG = "more info on replacements coming soon";

export const COLOR_MESSAGES = {
  background: `see ${CSS_DOCS_URL}/background-color to find suitable semantic replacement`,
  border: `see ${CSS_DOCS_URL}/border-color to find suitable semantic replacement`,
  text: `see ${CSS_DOCS_URL}/text-color to find suitable semantic replacement`,
};

export const TYPES = {
  replaced: "[REPLACED]",
  removed: "[REMOVED]"
};

const CLASS_MAP = {
  button: "use Warp button component instead",
  link: "use Warp button component instead",
  field: "use Warp form components instead",
  card: "use Warp card component instead",
  input: "use Warp input component instead",
  segment: "use Warp button-group (Vue) or Toggle component instead",
  expandable: "use Warp expandable component instead",
  modal: "use Warp modal component instead",
  slider: "use Warp slider component instead",
  thumb: "use Warp slider component instead",
  spinner: "use 'animate-spinner' instead",
  grid: "",
  step: "use Warp steps component instead",
  switch: "use Warp switch component instead",
  tab: "use Warp tabs component instead",
  toast: "use Warp toast component instead",
  list: "use Warp list component instead",
};

export const emitWarning = (selector, deprecationType, message) => {
  const className = Object.keys(CLASS_MAP).find(key => selector.includes(key));
  const warningMessage = !message && className ? CLASS_MAP[className] : message;
  console.warn(`${deprecationType === TYPES.removed ? CONSOLE_COLORS.error : CONSOLE_COLORS.warn}${deprecationType}${CONSOLE_COLORS.reset} ${selector}${warningMessage ? " -> " + warningMessage : ""}`);
};

export const colors = ["white", "black", "blue", "green", "aqua", "yellow", "red", "bluegray", "gray"];
export const colorHues = Array.from({ length: 10 }, (_, index) => index === 0 ? 50 : index * 100); // 50, 100, 200, ..., 900

export const COLOR_REGEXPS = {
  border: new RegExp("^border-(" + colors.join("|") + ")(-(\\d+))?$"),
  background: new RegExp("^bg-(" + colors.join("|") + ")(-(\\d+))?$"),
  text: new RegExp("^text-(" + colors.join("|") + ")(-(\\d+))?$"),
  divide: new RegExp("^divide-(" + colors.join("|") + ")(-(\\d+))?$"),
};
