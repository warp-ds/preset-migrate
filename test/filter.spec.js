import { describe, expect, test, vi, beforeEach } from "vitest";
import { setup } from "./_helpers.js";
import { TYPES } from "../utils.js"

setup();

const filterHues = [0, 50, 75, 90, 95, 100, 105, 110, 125, 150, 200];
const filterTypes = ['brightness', 'contrast', 'hue-rotate', 'saturate'];
const opacityHues= [0, 5, 10, 20, 25, 30, 40, 50, 60, 70, 75, 80, 90, 95, 100];
const blurDimensions = ['sm', 'md', 'lg', 'xl', '2xl', '3xl'];


describe('filter',  () => {
    let warnSpy;
    beforeEach(() => {
        warnSpy = vi.spyOn(global.console, 'warn');
    });

    test('prints a warning for using deprecated filters with static values', async ({ uno }) => {
        const classes = [
            'backdrop-filter',
            'backdrop-filter-none',
            'backdrop-invert-0',
            'invert-0',
            'backdrop-sepia-0',
            'sepia-0',
        ];
        const { css } = await uno.generate(classes);
        expect(css).toMatchInlineSnapshot('""');
        expect(warnSpy).toHaveBeenCalledTimes(classes.length);
        expect(warnSpy.calls.flat()).toMatchSnapshot();  
    });

    test('prints a warning for using deprecated filters brightness|contrast|hue-rotate|saturate', async ({ uno }) => {
        const classes = [...filterTypes.flatMap(t => filterHues.flatMap((v) => [`backdrop-${t}-${v}`, `${t}-${v}`]))];
        const { css } = await uno.generate(classes);
        expect(css).toMatchInlineSnapshot('""');
        expect(warnSpy).toHaveBeenCalledTimes(classes.length);
        expect(warnSpy.calls.flat()).toMatchSnapshot();  
    });

    test('prints a warning for using deprecated backdrop opacity filter', async ({ uno }) => {
        const classes = opacityHues.map(o => `backdrop-opacity-${o}`);
        const { css } = await uno.generate(classes);
        expect(css).toMatchInlineSnapshot('""');
        expect(warnSpy).toHaveBeenCalledTimes(classes.length);
        expect(warnSpy.calls.flat()).toMatchSnapshot();  
    });

    test('prints a warning for using deprecated blur filter', async ({ uno }) => {
        const classes = blurDimensions.map(o => `blur-${o}`);
        const { css } = await uno.generate(classes);
        expect(css).toMatchInlineSnapshot('""');
        expect(warnSpy).toHaveBeenCalledTimes(classes.length);
        expect(warnSpy.calls.flat()).toMatchSnapshot();  
    });

    test('prints a warning for using deprecated backdrop-blur filter', async ({ uno }) => {
        const classes = blurDimensions.map(o => `backdrop-blur-${o}`);
        const { css } = await uno.generate(classes);
        expect(css).toMatchInlineSnapshot('""');
        expect(warnSpy).toHaveBeenCalledTimes(classes.length-1);
        expect(warnSpy.calls.flat()).toMatchSnapshot();
    });
});
