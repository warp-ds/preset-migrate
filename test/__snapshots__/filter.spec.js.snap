// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filter > prints a warning for using deprecated (backdrop) blur filter 2`] = `
[
  "[91m[REMOVED][0m blur-sm",
  "[91m[REMOVED][0m backdrop-blur-sm",
  "[91m[REMOVED][0m blur-md",
  "[91m[REMOVED][0m backdrop-blur-md",
  "[91m[REMOVED][0m blur-lg",
  "[91m[REMOVED][0m backdrop-blur-lg",
  "[91m[REMOVED][0m blur-xl",
  "[91m[REMOVED][0m backdrop-blur-xl",
  "[91m[REMOVED][0m blur-2xl",
  "[91m[REMOVED][0m backdrop-blur-2xl",
  "[91m[REMOVED][0m blur-3xl",
  "[91m[REMOVED][0m backdrop-blur-3xl",
]
`;

exports[`filter > prints a warning for using deprecated backdrop opacity filter 2`] = `
[
  "[91m[REMOVED][0m backdrop-opacity-0",
  "[91m[REMOVED][0m backdrop-opacity-5",
  "[91m[REMOVED][0m backdrop-opacity-10",
  "[91m[REMOVED][0m backdrop-opacity-20",
  "[91m[REMOVED][0m backdrop-opacity-25",
  "[91m[REMOVED][0m backdrop-opacity-30",
  "[91m[REMOVED][0m backdrop-opacity-40",
  "[91m[REMOVED][0m backdrop-opacity-50",
  "[91m[REMOVED][0m backdrop-opacity-60",
  "[91m[REMOVED][0m backdrop-opacity-70",
  "[91m[REMOVED][0m backdrop-opacity-75",
  "[91m[REMOVED][0m backdrop-opacity-80",
  "[91m[REMOVED][0m backdrop-opacity-90",
  "[91m[REMOVED][0m backdrop-opacity-95",
  "[91m[REMOVED][0m backdrop-opacity-100",
]
`;

exports[`filter > prints a warning for using deprecated filters brightness|contrast|hue-rotate|saturate 2`] = `
[
  "[91m[REMOVED][0m backdrop-brightness-0",
  "[91m[REMOVED][0m brightness-0",
  "[91m[REMOVED][0m backdrop-brightness-50",
  "[91m[REMOVED][0m brightness-50",
  "[91m[REMOVED][0m backdrop-brightness-75",
  "[91m[REMOVED][0m brightness-75",
  "[91m[REMOVED][0m backdrop-brightness-90",
  "[91m[REMOVED][0m brightness-90",
  "[91m[REMOVED][0m backdrop-brightness-95",
  "[91m[REMOVED][0m brightness-95",
  "[91m[REMOVED][0m backdrop-brightness-100",
  "[91m[REMOVED][0m brightness-100",
  "[91m[REMOVED][0m backdrop-brightness-105",
  "[91m[REMOVED][0m brightness-105",
  "[91m[REMOVED][0m backdrop-brightness-110",
  "[91m[REMOVED][0m brightness-110",
  "[91m[REMOVED][0m backdrop-brightness-125",
  "[91m[REMOVED][0m brightness-125",
  "[91m[REMOVED][0m backdrop-brightness-150",
  "[91m[REMOVED][0m brightness-150",
  "[91m[REMOVED][0m backdrop-brightness-200",
  "[91m[REMOVED][0m brightness-200",
  "[91m[REMOVED][0m backdrop-contrast-0",
  "[91m[REMOVED][0m contrast-0",
  "[91m[REMOVED][0m backdrop-contrast-50",
  "[91m[REMOVED][0m contrast-50",
  "[91m[REMOVED][0m backdrop-contrast-75",
  "[91m[REMOVED][0m contrast-75",
  "[91m[REMOVED][0m backdrop-contrast-90",
  "[91m[REMOVED][0m contrast-90",
  "[91m[REMOVED][0m backdrop-contrast-95",
  "[91m[REMOVED][0m contrast-95",
  "[91m[REMOVED][0m backdrop-contrast-100",
  "[91m[REMOVED][0m contrast-100",
  "[91m[REMOVED][0m backdrop-contrast-105",
  "[91m[REMOVED][0m contrast-105",
  "[91m[REMOVED][0m backdrop-contrast-110",
  "[91m[REMOVED][0m contrast-110",
  "[91m[REMOVED][0m backdrop-contrast-125",
  "[91m[REMOVED][0m contrast-125",
  "[91m[REMOVED][0m backdrop-contrast-150",
  "[91m[REMOVED][0m contrast-150",
  "[91m[REMOVED][0m backdrop-contrast-200",
  "[91m[REMOVED][0m contrast-200",
  "[91m[REMOVED][0m backdrop-hue-rotate-0",
  "[91m[REMOVED][0m hue-rotate-0",
  "[91m[REMOVED][0m backdrop-hue-rotate-50",
  "[91m[REMOVED][0m hue-rotate-50",
  "[91m[REMOVED][0m backdrop-hue-rotate-75",
  "[91m[REMOVED][0m hue-rotate-75",
  "[91m[REMOVED][0m backdrop-hue-rotate-90",
  "[91m[REMOVED][0m hue-rotate-90",
  "[91m[REMOVED][0m backdrop-hue-rotate-95",
  "[91m[REMOVED][0m hue-rotate-95",
  "[91m[REMOVED][0m backdrop-hue-rotate-100",
  "[91m[REMOVED][0m hue-rotate-100",
  "[91m[REMOVED][0m backdrop-hue-rotate-105",
  "[91m[REMOVED][0m hue-rotate-105",
  "[91m[REMOVED][0m backdrop-hue-rotate-110",
  "[91m[REMOVED][0m hue-rotate-110",
  "[91m[REMOVED][0m backdrop-hue-rotate-125",
  "[91m[REMOVED][0m hue-rotate-125",
  "[91m[REMOVED][0m backdrop-hue-rotate-150",
  "[91m[REMOVED][0m hue-rotate-150",
  "[91m[REMOVED][0m backdrop-hue-rotate-200",
  "[91m[REMOVED][0m hue-rotate-200",
  "[91m[REMOVED][0m backdrop-saturate-0",
  "[91m[REMOVED][0m saturate-0",
  "[91m[REMOVED][0m backdrop-saturate-50",
  "[91m[REMOVED][0m saturate-50",
  "[91m[REMOVED][0m backdrop-saturate-75",
  "[91m[REMOVED][0m saturate-75",
  "[91m[REMOVED][0m backdrop-saturate-90",
  "[91m[REMOVED][0m saturate-90",
  "[91m[REMOVED][0m backdrop-saturate-95",
  "[91m[REMOVED][0m saturate-95",
  "[91m[REMOVED][0m backdrop-saturate-100",
  "[91m[REMOVED][0m saturate-100",
  "[91m[REMOVED][0m backdrop-saturate-105",
  "[91m[REMOVED][0m saturate-105",
  "[91m[REMOVED][0m backdrop-saturate-110",
  "[91m[REMOVED][0m saturate-110",
  "[91m[REMOVED][0m backdrop-saturate-125",
  "[91m[REMOVED][0m saturate-125",
  "[91m[REMOVED][0m backdrop-saturate-150",
  "[91m[REMOVED][0m saturate-150",
  "[91m[REMOVED][0m backdrop-saturate-200",
  "[91m[REMOVED][0m saturate-200",
]
`;

exports[`filter > prints a warning for using deprecated filters with static values 2`] = `
[
  "[91m[REMOVED][0m backdrop-blur",
  "[91m[REMOVED][0m backdrop-filter",
  "[91m[REMOVED][0m backdrop-filter-none",
  "[91m[REMOVED][0m backdrop-invert-0",
  "[91m[REMOVED][0m invert-0",
  "[91m[REMOVED][0m backdrop-sepia-0",
  "[91m[REMOVED][0m sepia-0",
]
`;
